# Feature Annotation

| Feature Type | _D. melanogaster_ | _D. simulans_ |
|--------:|:-------:|:------:|
| Genes | 17737 | 15385 |
| Transcripts | 35254 | 26261 |
| TSS (300bp Windows) | 22893 | 21069 |
| 5’UTR | 28479 | 25081 |
| 3’UTR | 21600 | 16231 |
| Exonic Features | 87473 | 79405 |
| Intronic Features | 44769 | 47236 |
| Intergenic Features | 11356 | 16174 |

** Supplementary Table 3 – Unique feature counts.** Counts of features within _D. melanogaster_ and _D. simulans_.  5’ UTR and 3’UTR were determined for each transcript using the references above.  A transcription start site (TSS) was defined as a 300 bp region, 150 bp upstream and downstream from each annotated transcript start.  In _D. melanogaster_ there were three pairs of genes that had the same start position, one on the forward strand and one on the reverse strand.  These included the gene pairs bug (FBgn0034050) and Diap2 (FBgn0015247), lncRNA:CR44456 (FBgn0265649) and lncRNA:CR44455 (FBgn0265648), and CR43482 (FBgn0263493) and CR43483 (FBgn0263494).  Event analysis (Newman, et al. 2018) was used to determine exonic and intronic features. Intergenic features were defined by subtracting the genic features from the entire genome with a length greater than 50 bp.
