# ChIP-seq Feature Analysis

## Supplementary Figure 6 – ChIP-seq feature FRiP.
![Supplementary Figure 6](https://github.com/McIntyre-Lab/papers/blob/master/nanni_chip_rna_2022/chipseq_features/mel_sim_summary_frip_feature_updated.png)
**Supplementary Figure 6 – ChIP-seq feature FRiP.**  Fraction of reads in each feature type for A. _D. melanogaster_ and B. _D. simulans_. FRiP scores are on the Y-axis and on the X-axis are the individual samples separated by H3K4me3, H3K27me2me3, or Input.  Each set contains 3 replicates of both male and female samples (N=6). A _D. simulans_ outlier sample (K4_sim_sz11_m_noEtoh_rep3) had higher immunoprecipitation specificity than other samples and was excluded from plot. Within each sample, the ratios of FRiP values between feature types is consistent.

&nbsp;
&nbsp;

## Supplementary Figure 7 – ChIP-seq Kappa Agreement
![Supplementary Figure 7](https://github.com/McIntyre-Lab/papers/blob/master/nanni_chip_rna_2022/chipseq_features/_kappa_bar_graph.png)
**Supplementary Figure 7 – ChIP-seq Kappa Agreement.**  Cohen’s Kappa agreement values (Fleiss 1981) across all feature types in _D. melanogaster_ and _D. simulans_. Kappa agreement corrects for marginal frequencies, so it is a more accurate measure compared to simple agreement.

&nbsp;
&nbsp;

## Supplementary Table 2 – Agreement of chromatin marks in orthologous genes.

| | Open Chromatin | Closed Chromatin |
|--------|:-------:|:------:|
| X Chromosome | 0.66 | 0.57 |
| Autosomes | 0.67 | 0.61 |

**Supplementary Table 2 – Agreement of chromatin marks in orthologous genes.** Cohen’s Kappa values (Fleiss 1981) indicating agreement of one-to-one orthologous genes with open chromatin marks (H3K4me3) or closed chromatin marks (H3K27me2me3) between _D. melanogaster_ and _D. simulans_ on the X chromosome and autosomes. Chromosome 4 is excluded from the autosomes.
