# Nanni, et al. 2022 Supplementary Information
### CCS long reads
- [What is the length distribution of CCS reads from each genotype?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_ccs_reads/ccs_readLength_distributions)
- [How well do CCS reads from each genotype map to B73 v4, Mo17 Yan, and Mo17 Cau reference genomes?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_ccs_reads/ccs_map_id_cov_counts)
- [What is the length distribution of CCS reads matching junction chains of B73 v4 annotated transcripts?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_ccs_reads/ccs_readLength_distributions)
### IsoSeq3 clusters
- [What is the quality of the cluster sequences for each genotype following IsoSeq3?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_clusters/cluster_fastqc_output)
- [How well do cluster sequences from each genotype map to B73 v4, Mo17 Yan, and Mo17 Cau reference genomes?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_clusters/cluster_map_id_cov_counts)
- [Do clusters from B73 and Mo17 genotypes map to both the B73 v4 and Mo17 Cau genomes in B73-Mo17 syntenic genes?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_clusters/compare_clusters_in_B73_Mo17_synteny)
- [What is the length distribution of clusters ignored by TOFU2 Cupcake due to low coverage/identity alignments to each of the three genomes?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_clusters/ignored_clusterLength_distributions)
- [Do clusters ignored by TOFU2 Cupcake due to low coverage/identity alignments to B73 v4 genome BLAST to known sequences?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_clusters/ignored_cluster_blast_results)
### Assembled transcripts
- [What is the length distribution of assembled transcripts mapped to B73 v4 genome following TOFU2 Cupcake collapse?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_assembled_transcripts/assembXcrpt_length_distributions)
- [What functional annotation (GO) are enriched in sets of differentially expressed genes?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_assembled_transcripts/DE_GO_results)
- [Are expression values similar within one-to-one B73-Mo17 syntenic gene pairs when mapped to the B73 v4 reference genome vs. Mo17 Cau?](https://github.com/McIntyre-Lab/papers/tree/master/nanni_maize_2022/suppl_assembled_transcripts/B73_vs_Mo17_CAU_mapped_read_scatter)

![Supplementary information in flowchart](https://raw.githubusercontent.com/McIntyre-Lab/papers/blob/master/nanni_maize_2022/master/nanni_maize_2022/supplement_github_flowchart.drawio.svg)

