#!/bin/bash

#SBATCH --mail-user=adalena.nanni@ufl.edu
#SBATCH --mail-type=FAIL
#SBATCH --job-name=create_inst
#SBATCH --output=/ufrc/mcintyre_conesa_transvar/isoAnnot/maize_ainsworth_PB/scripts/SLURM_LOGS/combine_annotations_%A.out
#SBATCH --time=01:00:00
#SBATCH --ntasks=1
#SBATCH --mem=1gb
#SBATCH --cpus-per-task=1

module purge
module load java/1.8.0_31 python perl

### Parse of InterProScan
### Combine all related parsed files into one
### 	(all files are still split)


### Set Directories
PROJ=/ufrc/mcintyre_conesa_transvar/isoAnnot/maize_ainsworth_PB
SCRIPT=/ufrc/mcintyre_conesa_transvar/isoAnnot/sorghum_maize_analysis/scripts
PACBIO=/ufrc/mcintyre/share/maize_ainsworth/sqanti_post_filter
OUT=$PROJ/generate_anno
    mkdir -p ${OUT}

SPECIES=maize

## Set Instructions File Name
INST=$OUT/${SPECIES}.instructions4tappas.cfg


###############    
#### INPUT ####
###############

## Get Pacbio GTF File
PACBIOGTF=${PACBIO}/sqanti_filtered_corrected.gtf

## Gene ontology (GO) annotation
#GO=$PROJ/blast_2_go/${SPECIES}_FSM_ISM.blast2go.tsv

## Get SQANTI Classification and Junction Files
CLASS=${PACBIO}/SQANTI_classification.txt
JUNC=${PACBIO}/SQANTI_junctions.txt

## Get Scan For Motifs UTR Scan Annotation Files
SFM=$PROJ/scanformotif/${SPECIES}.??.SFM.out

## Get miRNA Annotation
MIRNA=$PROJ/miRNA_target/parsed_psRNATarget.txt

## Get Final InterproScan Output TSV File
IPRSCAN=$PROJ/interproscan/final.${SPECIES}_interproscan_out.tsv

## Nonsense Mediated Decay From Protein Association
NMD=$PROJ/protein_db_association/total_${SPECIES}_corrected_NMD.txt

## Protein Association Output From Protein	Association
PROTASSO=$PROJ/protein_db_association/total_${SPECIES}_corrected_proteinAssociation.txt


################
#### OUTPUT ####
################

## Final Annotation Output
ANNO=${OUT}/${SPECIES}.annotation_output4TAPPAS.gff

echo '#!/bin/sh' > ${INST}
echo "
## ISOannot for PacBioMix2017 (pacbio + some refseq)
## Lorena de la Fuente
## 27/10/2016

## Generated by create_instructions_cfg.sbatch on $(date)


#####################################
####### PacBio information #########
#####################################

db=\"pacbio\"

# GTF file

gtf_file=\"${PACBIOGTF}\"

######################################################################
######## Generic information by creating a sqanti like output ########
######################################################################

# sqanti_qc.py

output_class=\"${CLASS}\"
output_junction=\"${JUNC}\"

#gene level: name or id
gene_level=\"name\"

# referenceProtToPacbioProt.py 

output_NMD=\"${NMD}\"
protein_association=\"${PROTASSO}\"

####################################
####### Gene Level Annotation ######
####################################

#gene_annot=\"${GO}\"

#################################
####### UTRSCAN_annotation ######
#################################

" >> $INST

for file in ${SFM} ; do echo "output_scanForMotifs=\"$file\""; done  >> ${INST}

echo "

###############################
####### miRNA annotation ######
###############################


output_miRNA=\"${MIRNA}\"


####################################################################
####### InterproScan data: PFAM + SIGNALP + TMHMM + MobiDBLite #####
#####################################################################

output_interproscan=\"${IPRSCAN}\"


#############################################################
######### CREATING ANNOTATION FILE FROM ABOVE FILES #########
#############################################################

annotation_output=\"${ANNO}\"

"  >> ${INST}

## Run Annotation Generation Script
module load python

python ${SCRIPT}/t2gAnnotationFilePacbio2017.py ${INST}


## Adding in Gene Ontology lines
## The t2gAnnotationFilePacbio2017.py script does not recognize
## 	Blast2GO output (PB isoform in 1st column instead of accession)

#awk -F '\t' '{print $1"\t"$3"\t"$4"\t.\t.\t.\t.\t.\tID="$2"; Name="$2"; Desc="$5}'\
#	${GO} >> ${ANNO}

## Replace all T2GO elements in the 2nd column with tappAS and sort
## Sort MUST be stable so that the gff elements stay in the correct order
sed -i 's/T2GO/tappAS/g' ${ANNO}
sort -k1,1 -s ${ANNO} > $(dirname ${ANNO})/final_$(basename ${ANNO})

## Remove intermediate annotation file (before reformat)
rm ${ANNO}
